swift/ql/lib/codeql/swift/elements/expr/Expr.qll

```diff
@@ -1,4 +1,17 @@
-// generated by codegen/codegen.py, remove this comment if you wish to edit this file
 private import codeql.swift.generated.expr.Expr
 
-class Expr extends ExprBase { }
+class Expr extends ExprBase {
+  final override Expr getResolveStep() { convertsFrom(result) }
```

Comment:
I don't really understand what's going on with the whole resolve mechanism - is it just the inverse of `getConverted` or is there more to it?

Reply:
That's pretty much it, yes. AST classes can override `getResolveStep` to not be included in the main AST (i.e., when navigating the parent/child nodes in the AST classes). This is done by autogenerating AST classes lik:
```ql
class FooBase {
  Element getBar() {
    exists(Bar bar |
      bar_of_foo(this, bar) and
      result = bar.resolve()
    )
  }
}
```
where `resolve` is then the TC of `getResolveStep`.

Reply:
Currently, `resolve` only does something in the case of `IdentityExprs` (which in particular include parenthesis expressions), `ExplicitCast`s and `ImplicitConversion`s. For the `Expr` classes we provide aliases for `result = getResolveStep()` as `convertsFrom(result)` (the result became a parameter because we couldn't think of a good name), and `getConversion(x)` can then be implemented as "whatever expression is the result that converted something from `x`" (i.e., `Expr getConversion() { result.convertsFrom(this) }`).

Does that make sense?